jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java

```diff
@@ -0,0 +1,304 @@
+    {
+        this(byteBufferPool, (ByteChannel)seekableByteChannel, offset, length);
+        if (offset >= 0 && seekableByteChannel != null)
+        {
+            try
+            {
+                seekableByteChannel.position(offset);
+            }
+            catch (IOException e)
+            {
+                _terminal = Content.Chunk.from(e, true);
+            }
+        }
+    }
+
+    public ByteChannelContentSource(ByteChannel byteChannel)
+    {
+        this(new ByteBufferPool.Sized(null), byteChannel, -1L, -1L);
+    }
+
+    public ByteChannelContentSource(ByteBufferPool.Sized byteBufferPool, ByteChannel byteChannel)
+    {
+        this(byteBufferPool, byteChannel, -1L, -1L);
+    }
+
+    private ByteChannelContentSource(ByteBufferPool.Sized byteBufferPool, ByteChannel byteChannel, long offset, long length)
+    {
+        _byteBufferPool = Objects.requireNonNull(byteBufferPool);
+        _byteChannel = byteChannel;
+        _offset = offset < 0 ? 0 : offset;
+        _length = length;
+    }
+
+    protected ByteChannel open() throws IOException
+    {
+        return _byteChannel;
+    }
+
+    @Override
+    public void demand(Runnable demandCallback)
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            if (this.demandCallback != null)
+                throw new IllegalStateException("demand pending");
+            this.demandCallback = demandCallback;
+        }
+        _invoker.run(this::invokeDemandCallback);
+    }
+
+    private void invokeDemandCallback()
+    {
+        Runnable demandCallback;
+        try (AutoLock ignored = lock.lock())
+        {
+            demandCallback = this.demandCallback;
+            this.demandCallback = null;
+        }
+        if (demandCallback != null)
+            ExceptionUtil.run(demandCallback, this::fail);
+    }
+
+    private void lockedCheckOpen()
+    {
+        if (_terminal == null || !Content.Chunk.isFailure(_terminal) && _terminal.isLast())
+        {
+            _terminal = null;
+            if (_byteChannel == null || !_byteChannel.isOpen())
+            {
+                try
+                {
+                    _byteChannel = open();
+                    if (_byteChannel == null || !_byteChannel.isOpen())
+                        _terminal = Content.Chunk.from(new ClosedChannelException(), true);
+                    if (_offset > 0 && _byteChannel instanceof SeekableByteChannel seekableByteChannel)
+                        seekableByteChannel.position(_offset);
+                }
+                catch (IOException e)
+                {
+                    _terminal = Content.Chunk.from(e, true);
+                }
+
+                if (_terminal != null && _buffer != null)
+                {
+                    _buffer.release();
+                    _buffer = null;
+                }
+            }
+        }
+    }
```

Comment:
I think this method does way too much, too many side effects calling it.

I think there should be a method used from `read()` that opens the channel if possible, and a different method to be used from `rewind()`.

For example right now: app reads, then reads EOF; if the app insist reading it, it should receive another EOF, but instead `lockedCheckOpen()` will clear _terminal and re-open the channel -- not good.

I feel this class should inherit the simplicity of `PathContentSource` rather than the current more complex implementation.

Reply:
Well this class was simple, but @lorban asked for it to be used to replace many other similar classes, so it is now a bit more complex.   Ultimately this class should be able to replace 3 or 4 other classes once some small behaviour changes are permitted in 12.1.

However, I will remove the rewind logic from this method.